dags/mlflow-dag.py

```diff
@@ -0,0 +1,239 @@
+By using an external XCom backend, users can easily push and pull all intermediary data generated in their DAG in GCS.
+"""
+
+
+def log_roc_curve(y_test: list, y_pred: list):
+    fpr, tpr, thresholds = roc_curve(y_test, y_pred)
+    plt.plot(fpr,tpr) 
+    plt.ylabel('False Positive Rate')
+    plt.xlabel('True Positive Rate')
+    plt.title('ROC Curve')
+    plt.savefig("roc_curve.png")
+    mlflow.log_artifact("roc_curve.png")
+    plt.close()
+
+
+def log_confusion_matrix(y_test: list, y_pred: list):
+    cm = confusion_matrix(y_test, y_pred)
+    t_n, f_p, f_n, t_p = cm.ravel()
+    mlflow.log_metrics({'True Positive': t_p, 'True Negative': t_n, 'False Positive': f_p, 'False Negatives': f_n})
+
+    ConfusionMatrixDisplay.from_predictions(y_test, y_pred)
+    plt.savefig("confusion_matrix.png")
+    mlflow.log_artifact("confusion_matrix.png")
+    plt.close()
+
+
+def log_classification_report(y_test: list, y_pred: list):
+    cr = classification_report(y_test, y_pred, output_dict=True)
+    logging.info(cr)
+    cr_metrics = pd.json_normalize(cr, sep='_').to_dict(orient='records')[0]
+    mlflow.log_metrics(cr_metrics)
+
+
+def log_all_eval_metrics(y_test: list, y_pred: list):
+    
+    # Classification Report
+    log_classification_report(y_test, y_pred)
+
+
+    # Confustion Matrix
+    log_confusion_matrix(y_test, y_pred)
+
+
+    # ROC Curve
+    log_roc_curve(y_test, y_pred)
+
+
+@dag(
+    start_date=datetime(2021, 1, 1),
+    schedule_interval=None,
+    catchup=False,
+    doc_md=docs
+)
+def mlflow_example():
+
+    @task
+    def load_data():
+        """Pull Census data from Public BigQuery and save as Pandas dataframe in GCS bucket with XCom"""
+
+        bq = BigQueryHook()
+        sql = """
+        SELECT * FROM `bigquery-public-data.ml_datasets.census_adult_income`
+        """
+
+        return bq.get_pandas_df(sql=sql, dialect='standard')
+
+
+
+    @task
+    def preprocessing(df: pd.DataFrame):
+        """Clean Data and prepare for feature engineering
+        
+        Returns pandas dataframe via Xcom to GCS bucket.
+
+        Keyword arguments:
+        df -- Raw data pulled from BigQuery to be processed. 
+        """
+
+        df.dropna(inplace=True)
+        df.drop_duplicates(inplace=True)
+
+        # Clean Categorical Variables (strings)
+        cols = df.columns
+        for col in cols:
+            if df.dtypes[col]=='object':
+                df[col] =df[col].apply(lambda x: x.rstrip().lstrip())
+
+
+        # Rename up '?' values as 'Unknown'
+        df['workclass'] = df['workclass'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['occupation'] = df['occupation'].apply(lambda x: 'Unknown' if x == '?' else x)
+        df['native_country'] = df['native_country'].apply(lambda x: 'Unknown' if x == '?' else x)
+
+
+        # Drop Extra/Unused Columns
+        df.drop(columns=['education_num', 'relationship', 'functional_weight'], inplace=True)
```

Comment:
I find it more helpful to keep what I want to keep rather than drop what I don't want. It forces one to know the schema and prevents surprises down the line.